pkg/kubelet/server/stats/fs_resource_analyzer.go

```diff
@@ -0,0 +1,146 @@
+	}
+}
+
+// Start eager background caching of volume stats.
+func (s *fsResourceAnalyzer) Start() {
+	go func() {
+		sleepTime := s.calcVolumeFrequency
+		if sleepTime <= 0 {
+			glog.Warningf("volume stats calc interval must be greater than 0, defaulting to %v.", defaultVolumeStatsCalcInterval)
+			sleepTime = defaultVolumeStatsCalcInterval
+		}
+		for {
+			glog.V(3).Infof("Calculating volume stats and caching result.")
+			startTime := time.Now()
+			s.updateCachedPodVolumeStats()
+			elapsedDuration := time.Now().Sub(startTime)
+			glog.V(3).Infof("Finished calculating volume stats in %v, sleeping for %v.", elapsedDuration, sleepTime)
+			time.Sleep(sleepTime)
+		}
+	}()
+}
+
+// updateCachedPodVolumeStats calculates and caches the PodVolumeStats for every Pod known to the kubelet.
+func (s *fsResourceAnalyzer) updateCachedPodVolumeStats() {
+	// Calculate the new volume stats map
+	pods := s.statsProvider.GetPods()
+	newCache := &cachedPodVolumeStats{make(map[types.UID]*PodVolumeStats, len(pods))}
+	for _, pod := range pods {
+		podUid := pod.GetUID()
+		stats, err := s.getPodVolumeStats(podUid)
+		if err != nil {
+			continue
+		}
+		newCache.volumeStats[podUid] = &stats
+	}
+
+	// Update the map reference
+	s.volumeCache.Lock()
+	defer s.volumeCache.Unlock()
+	s.cachedVolumeStats = newCache
+}
+
+// getPodVolumeStats calculates PodVolumeStats for a given pod and returns the result.
+func (s *fsResourceAnalyzer) getPodVolumeStats(uid types.UID) (PodVolumeStats, error) {
+	// Find all Volumes for the Pod
+	volumes, err := s.statsProvider.ListVolumesForPod(uid)
+	if err != nil {
+		return PodVolumeStats{}, err
+	}
+
+	// Call GetMetrics on each Volume and copy the result to a new VolumeStats.FsStats
+	stats := make([]VolumeStats, 0, len(volumes))
+	for name, v := range volumes {
+		metric, err := v.GetMetrics()
+		if err != nil {
+			// Expected for Volumes that don't support Metrics
+			// TODO: Disambiguate unsupported from errors
+			glog.V(3).Infof("Failed to calculate volume metrics for pod %v volume %s: %v", uid, name, err)
+			continue
+		}
+		stats = append(stats, s.parsePodVolumeStats(name, metric))
+	}
+	return PodVolumeStats{PodUid: uid, Volumes: stats}, nil
+}
+
+func (s *fsResourceAnalyzer) parsePodVolumeStats(podName string, metric *volume.Metrics) VolumeStats {
+	return VolumeStats{
+		Name: podName,
+		FsStats: FsStats{
+			AvailableBytes: metric.Available,
+			CapacityBytes:  metric.Capacity,
+			UsedBytes:      metric.Used}}
+}
+
+// GetPodVolumeStats returns the PodVolumeStats for a given pod.  Results are looked up from a cache that
+// is eagerly populated in the background, and never calculated on the fly.
+func (s *fsResourceAnalyzer) GetPodVolumeStats(uid types.UID) (PodVolumeStats, error) {
+	s.volumeCache.RLock()
+	defer s.volumeCache.RUnlock()
+	stats, f := s.cachedVolumeStats.volumeStats[uid]
+	if !f {
+		return PodVolumeStats{}, fmt.Errorf("PodVolumeStats for pod %v not found.", uid)
```

Comment:
Somewhere in the call stack, the boolean has to be translated into a human friendly error message though.